### PR TITLE
feat: first implementation of policy gradient 

### PR DESCRIPTION
**Pull request by ThomasMousseau**
I experimented with solving the toy environment using policy gradient RL instead of GFlowNets. The goal was to keep it as minimal and close to the original GFlowNet implementation as possible. If there is any error or improvement, feel free to let me know!

Key differences:
- Uses policy gradient for training instead of GFlowNet.
- Runs batches of episodes (instead of single episodes) to reduce variance.

Results:
- As expected, the evaluation metric (MAE) is higher than with GFlowNets, since RL optimizes for maximizing reward rather than yielding samples proportional to reward.

Next steps:
- With Thursday’s quiz looming, any possibility of salvaging participation points will be taken. 
- I’d also be open to implementing this environment in JAX for the remaining points!

### CODE REVIEW

**Comment by ThomasMousseau** [comment]
@alexhernandezgarcia

File: rl.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        x = self.linear(x)
+        return x
+
+policy = Policy(n_states, float_type, device)
+
+### OPTIMIZER ###
+
+n_train_steps = 2000
+learning_rate = 0.01
+optimizer = torch.optim.Adam(policy.parameters(), lr=learning_rate)
```
Comment (alexhernandezgarcia) [inline rl.py]: I would use the same optimizer (SGD with momentum) as in the GFN implementation, for more direct comparison.

**Comment by alexhernandezgarcia** [comment]
@ThomasMousseau thank you for the very quick reaction to this! 

I still need to do a more in-depth check of the code, but my most immediate comment would be about training batches of episodes. I understand that this makes training more effective than updating the model with one episode at a time. It would be the same for GFlowNet. However, I decided to do it that way to minimise the complexity of the code and allow a reader of the code to understand things as easily as possible. The goal was not to make it efficient or competitive.

So questions:
- Did it not work well at all with just one episode at a time?
- Would you be willing to make the "batch size" equal to 1? We could still keep this version in a directory containing alternatives or improvements of the methods, if wanted.

PS: about JAX: I don't have experience with JAX and I'm very curious. Especially, I've heard that it could also improve the efficiency of GFlowNets training. So yes, you're very welcome to do that too and I'd be happy to consider that as course work and trade it for grade points :) In fact, if you wanna do something at a larger scale than one script, it could even become the course project. We can talk about it if you want.

**Comment by ThomasMousseau** [comment]
> @ThomasMousseau thank you for the very quick reaction to this!
> 
> I still need to do a more in-depth check of the code, but my most immediate comment would be about training batches of episodes. I understand that this makes training more effective than updating the model with one episode at a time. It would be the same for GFlowNet. However, I decided to do it that way to minimise the complexity of the code and allow a reader of the code to understand things as easily as possible. The goal was not to make it efficient or competitive.
> 
> So questions:
> 
> * Did it not work well at all with just one episode at a time?
> * Would you be willing to make the "batch size" equal to 1? We could still keep this version in a directory containing alternatives or improvements of the methods, if wanted.
> 
> PS: about JAX: I don't have experience with JAX and I'm very curious. Especially, I've heard that it could also improve the efficiency of GFlowNets training. So yes, you're very welcome to do that too and I'd be happy to consider that as course work and trade it for grade points :) In fact, if you wanna do something at a larger scale than one script, it could even become the course project. We can talk about it if you want.

Both the change of optimizer and the implementation of batch size were decisions purely made to follow best practices of model-free reinforcement learning and based on my previous projects. This was an error on my part, as I overlooked the comparison aspect of the exercise and got too focused on making it sample efficient.

Regarding the JAX part, I would be very interested in making this a project for the course. Let’s talk about it tomorrow!

**Review by alexhernandezgarcia — CHANGES_REQUESTED** [review]
Thanks for the changes!

All good with the optimiser.

About the batch size, sorry to be nitpicky with it, but would you be open to changing not just the actual number in the `batch_size` variable, but actually simplifying the code to actually remove the possibility of having a mini-batch of trajectories. The reason is that this is the way to remove this bit of additional complexity, and this is how the GFN version is implemented.

In practice, this would mean removing the batch for loops and the variables that are introduced to handle mini-batches.

Like I said, I'm open to keep this version somewhere else in the repo.

**Comment by ThomasMousseau** [comment]
> Thanks for the changes!
> 
> All good with the optimiser.
> 
> About the batch size, sorry to be nitpicky with it, but would you be open to changing not just the actual number in the `batch_size` variable, but actually simplifying the code to actually remove the possibility of having a mini-batch of trajectories. The reason is that this is the way to remove this bit of additional complexity, and this is how the GFN version is implemented.
> 
> In practice, this would mean removing the batch for loops and the variables that are introduced to handle mini-batches.
> 
> Like I said, I'm open to keep this version somewhere else in the repo.

No worries, I removed the batch_size variable and everything related to batch processing.
I also simplified the loss function to more closely follow the REINFORCE framework for simplicity purposes.
Additionally, I added comments on lines that might be less intuitive for newcomers to reinforcement learning.

File: rl.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+This script trains a minimal REINFORCE agent on the "Toy environment", which is the
+defined in Figure 2 of the GFlowNet Foundations paper, Bengio et al (JMLR, 2023):
+
+.. _a link: https://jmlr.org/papers/v24/22-0364.html
+"""
+
+"""
+Key differences with gfn.py:
+- Uses REINFORCE framework instead of GFlowNet training
+- Batches of episodes instead of single episodes to reduce variance
```
Comment (alexhernandezgarcia) [inline rl.py]: I think this is obsolete.

**Review by alexhernandezgarcia — APPROVED** [review]
Thanks for the changes. It looks good to me now! There's only an additional comment about something that's become obsolete. Happy to merge after that is updated.

**Review by alexhernandezgarcia — APPROVED** [review]
